torch/_dynamo/eval_frame.py

```diff
@@ -789,6 +793,75 @@ def run_node(self, n):
                 r.node.meta["val"] = self.current_node.meta["val"]
             return r
 
+    class AddRuntimeAssertsInInputConstraint(torch.fx.interpreter.Transformer):
```

Comment:
Hmm, I am still having some high level conceptual problems with this.

1. Does it really need to be fused into this function? This seems like something that would be perfectly happy to live as its own, separate transformation pass. You don't have to insert these runtime asserts, right? Maybe the input constraints are being checked some other way.
2. Are you sure these constraints are even what you want? The value ranges will typically look like [2, inf], but remember that 0/1 specialization says that we are unsoundly hoping the graph also works for size 1. You don't have any tests printing the generated asserts so I can't tell if you are generating an assert that's too strict.
3. The asserts, as generated, are very easy to accidentally DCE. Are we producing a functionalized graph or not here?

Reply:
1) We don't need to fuse it with this function. I just put it here as there was no other place to put it :) 
2) I think once the @avikchaudhuri 's work on surfacing guards and having more constrained input constraints, they can be used as asserts in the runtime. For now, they are just asserting the value is from 2 to inf. I like runtime asserts as they are very explicit about what we are protecting on. 
3) This PR doesn't produce functionalized graph. But, I am experimenting with the next PR to use AOTAutograd functionalization. So far, I was able to get following graph:
```
# source code
class Foo(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.register_buffer("buffer1", torch.ones(6, 1))
        self.register_buffer("buffer2", torch.ones(6, 2))

    def forward(self, x, y):
        self.buffer1.add_(2)
        x.add_(3)
        assert x.shape[0] > 4
        return x.sum() + self.buffer1.sum()


# resulting graph
def forward(self, arg1_1, arg2_1):
    _tensor_constant0 = self._tensor_constant0
    add = torch.ops.aten.add.Tensor(_tensor_constant0, 2);  _tensor_constant0 = None
    add_1 = torch.ops.aten.add.Tensor(arg1_1, 3)
    sym_size = torch.ops.aten.sym_size(arg1_1, 0);  arg1_1 = None
    gt = sym_size > 4;  sym_size = None
    scalar_tensor = torch.ops.aten.scalar_tensor.default(gt, device = device(type='cpu'), pin_memory = False);  gt = None
    _assert_async = torch.ops.aten._assert_async.msg(scalar_tensor, 'assertion error');  scalar_tensor = None
    sum_1 = torch.ops.aten.sum.default(add_1)
    sum_2 = torch.ops.aten.sum.default(add)
    add_2 = torch.ops.aten.add.Tensor(sum_1, sum_2);  sum_1 = sum_2 = None
    return (add, add_1, add_2)
 ```
Note that here we are asserting on the old value of input argument which can be unsound. So i was thinking about just replacing the returned output in the assert statement directly.

Reply:
Actually this might be problematic. It doesn't seem that easy to just reuse the output directly in the assert because we would then need to keep track whether assert comes before the inplace mutation or not.

Reply:
Just so I'm following, the above code is incorrect because `sym_size` should depend on `add_1` instead of `arg1_1`, right?

Reply:
yeah well it is kind of correct in the sense that assert will still work. For shape metadata mutation cases, it seems to capture the correct shape to assert. I am trying to verify this hypothesis now.

Reply:
> For now, they are just asserting the value is from 2 to inf. I like runtime asserts as they are very explicit about what we are protecting on.

What I am saying is, you probably don't want a 2 to inf bound. Because the agreement we had when we decided to do unsound 0/1 spec because we were going to guess that the graph would work anyway with size 1. If you burn in the assert here, then the model will definitely not work with size 1!

Reply:
Ah yeah good point! These asserts are not very useful until we have an API that users can specify a range